### PR TITLE
Add incremental pylint make target

### PR DESCRIPTION
**Pull request by levbishop**
Pylint is slow enough that seems like many contributors leave it to CI, which is an awkward way of working and ties up CI resources. This PR adds an `lint-incr` target that only runs pylint on files that are different on this branch and additionally turns on pylint parallelism. This can miss a few issues (especially cyclic-import) but its better than not running pylint at all and is a huge speedup for typical PRs that only touch a few files.

This works by updating a local reference to track HEAD of upstream github, and using this ref as merge-base for finding the files that are changed on the local branch. If working offline, this ref won't be updated, but this will merely have the effect that the merge-base may be further into the past than it could be (eg if the changes have been from main into the local branch since it was created) and more files get pylinted than is strictly necessary. This is only a performance issue, and either way is faster than the current approach of always pylinting all files.

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
I like this idea as a short term workaround. I do wonder if it's worth giving flake8 a harder look now because it doesn't have these same speed issues).

Can you add this to the tox.ini too (and replace `lint` in the default envlist so it gets run by default).

I do wonder about the origin remote default though I'm not sure how everyone has their local repo configured. It works fine for me where I use origin as the upstream and have a separate remote for my fork. But I think others might do something different.

**Comment by levbishop** [comment]
> I like this idea as a short term workaround. I do wonder if it's worth giving flake8 a harder look now because it doesn't have these same speed issues).

Yes I think we should investigate other tools such as flake8.

> Can you add this to the tox.ini too (and replace `lint` in the default envlist so it gets run by default).

Will do.

> I do wonder about the origin remote default though I'm not sure how everyone has their local repo configured. It works fine for me where I use origin as the upstream and have a separate remote for my fork. But I think others might do something different.

Agreed, can you think of a better solution?

- Some kind of `git merge-base`? Depends how the branch got started/rebased/etc.
- Parsing `git remote -v show -n` to find the remote whose url is qiskit github? Seems messy.
- Add a new remote temporarily pointing to qiskit github and then remove it after? Requires network access, but maybe this is good anyway since even if `origin/main` is the right branch it may not be up-to-date.

**Comment by levbishop** [comment]
Ok the current version should be robust against the users local git configuration.

Downsides of this approach:
1. Requires network access. Could maybe cache merge-base for times without network but if lint is a just-before-pushing thing then it isn't necessary. 
2. Requires a Unix shell for tox. Could make a short Python tools/pylint-incr.py to make it more cross platform

**Comment by levbishop** [comment]
Latest version opportunistically updates a local ref `refs/heads/lint_incr_latest` to the latest from github, but won't fail on lack of network access so long as the local ref has been set once.

Still requires unix shell for tox.

**Comment by levbishop** [comment]
I think I'm done messing with this. Noticed a couple other issues that are orthogonal to incremental pylint and I'll open separate PRs for them. Now works gracefully offline, and should also support windows (not tested). Added a para to contributing.md.

**Review by mtreinish — APPROVED** [review]
This lgtm, I like the solution you settled on with `git diff-index` and allowing the fetch to be ignored. I think this should be robust enough to work in every env (except for one where they're running lint on tarball of the repo without git (or a system without access to github at all). But in those situations a full lint run will still work fine.

File: tox.ini [hunk 1]
```diff
@@ -24,6 +24,20 @@ basepython = python3
...
+  reno lint
+
+[testenv:lint-incr]
+envdir = .tox/lint
+basepython = python3
+allowlist_externals = git
+commands =
+  black --check {posargs} qiskit test tools examples setup.py
+	-git fetch -q https://github.com/Qiskit/qiskit-terra.git :lint_incr_latest
+  {toxinidir}/tools/pylint_incr.py -rn -j4 -sn --paths :/qiskit/*.py :/test/*.py
```
[the remark this answers is not in the document]
Comment (levbishop) [reply tox.ini:39]: Yes, I'll update

File: tox.ini [hunk 2]
```diff
@@ -24,6 +24,20 @@ basepython = python3
 commands =
   black --check {posargs} qiskit test tools examples setup.py
   pylint -rn qiskit test tools
+  # pylint -rn --disable='invalid-name,missing-module-docstring,redefined-outer-name' examples/python/*.py
```
Comment (mtreinish) [inline tox.ini:27]:
```suggestion
  pylint -rn --disable='invalid-name,missing-module-docstring,redefined-outer-name' examples/python/*.py
```

**Review by mtreinish — DISMISSED** [review]

File: tox.ini (hunk 2 above)
Comment (levbishop) [reply tox.ini]: tox doesn't do globbing so this will fail....

**Review by mtreinish — APPROVED** [review]

**Pull request merged**